pkg/block/blocktest/multipart.go

```diff
@@ -0,0 +1,302 @@
+
+	_, err = adapter.CompleteMultiPartUpload(ctx, obj, resp.UploadID, &block.MultipartUploadCompletion{
+		Part: multiParts,
+	})
+	require.NoError(t, err)
+}
+
+// List parts when there are none
+func verifyListInvalid(t *testing.T, ctx context.Context, adapter block.Adapter, obj block.ObjectPointer, uploadID string) {
+	t.Helper()
+	_, err := adapter.ListParts(ctx, obj, uploadID, block.ListPartsOpts{})
+	if adapter.BlockstoreType() != block.BlockstoreTypeS3 {
+		require.ErrorIs(t, err, block.ErrOperationNotSupported)
+	} else {
+		require.NotNil(t, err)
+	}
+}
+
+// Upload parts after starting a multipart upload
+func uploadParts(t *testing.T, ctx context.Context, adapter block.Adapter, obj block.ObjectPointer, uploadID string, parts [][]byte) []block.MultipartPart {
+	t.Helper()
+	multiParts := make([]block.MultipartPart, len(parts))
+	for i, content := range parts {
+		partNumber := i + 1
+		partResp, err := adapter.UploadPart(ctx, obj, int64(len(content)), bytes.NewReader(content), uploadID, partNumber)
+		require.NoError(t, err)
+		multiParts[i].PartNumber = partNumber
+		multiParts[i].ETag = partResp.ETag
+	}
+
+	return multiParts
+}
+
+// Copy parts after starting a multipart upload
+func copyPartRange(t *testing.T, ctx context.Context, adapter block.Adapter, obj, objCopy block.ObjectPointer, uploadID string) []block.MultipartPart {
+	t.Helper()
+	multiParts := make([]block.MultipartPart, multipartNumberOfParts)
+	var startPosition int64 = 0
+	for i := 0; i < multipartNumberOfParts; i++ {
+		partNumber := i + 1
+		var endPosition = startPosition + multipartPartSize
+		partResp, err := adapter.UploadCopyPartRange(ctx, obj, objCopy, uploadID, partNumber, startPosition, endPosition)
+		if adapter.BlockstoreType() == block.BlockstoreTypeAzure {
+			require.ErrorContains(t, err, "not implemented") // azurite block store emulator did not yet implement this
+			return nil
+		}
+		require.NoError(t, err)
+		multiParts[i].PartNumber = partNumber
+		multiParts[i].ETag = partResp.ETag
+		startPosition = endPosition + 1
+	}
+
+	return multiParts
+}
+
+// Copy one and only part after starting a multipart upload
+func copyPart(t *testing.T, ctx context.Context, adapter block.Adapter, obj, objCopy block.ObjectPointer, uploadID string) ([]block.MultipartPart, error) {
+	t.Helper()
+	multiParts := make([]block.MultipartPart, 1)
+	partNumber := 1
+	partResp, err := adapter.UploadCopyPart(ctx, obj, objCopy, uploadID, partNumber)
+	if err != nil {
+		return nil, err
+	}
+	multiParts[0].PartNumber = partNumber
+	multiParts[0].ETag = partResp.ETag
+
+	return multiParts, nil
+}
+
+// getAndCheckContents GETs the Object from the Store, then compares its contents to the exp byte array
+func getAndCheckContents(t *testing.T, ctx context.Context, adapter block.Adapter, exp []byte, obj block.ObjectPointer) {
+	t.Helper()
+	// first check exists
+	ok, err := adapter.Exists(ctx, obj)
+	require.NoError(t, err)
```

Comment:
I'd also add custom messages here (like I explained above), as this can fail in multiple assertions.